### DISCUSSION TITLE
Referencing a local composite repo fails the build  if *any* of the proxy repos are missing/not yet built

### DISCUSSION THREAD

**Discussion by grosenberg** [eclipse-m2e/m2e-core#1312] [Q&A]
This setup was working prior to the update to the M2E 2.0.3.20230228-2348 version.

Question:
Is there a way to tell M2E not to fail a build immediately if *any* repo referenced in a composite repo is missing? That is, reduce the blocking ERROR to a Warning or Info that will allow the build to proceed?

The present requirement that *all* repos referenced in a composite repo be fully present effectually prevents building any.

Background:
I use a local composite repo to provides indirect inclusion of multiple local eclipse site repositories, some of which are feature included in other plugin projects and some that are wholly independent. There are no circular dependencies.

The parent POMs (of projects that include any other plugin) have a `<repository>` reference to the composite repo. (Highly convenient as there is now only a single location that contains all of the absolute paths to the included repos.)

After the update, the build of any project that references the composite repo fails if *any* repo indirectly referenced has not been built *regardless* of the fact that the missing repo is not actually needed.

>[INFO] Adding repository file:/D:/DevFiles/Eclipse/Update
[ERROR] [004b57a8-375c-4741-bd3b-95227a83d958][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.1] Provisioning exception
org.eclipse.equinox.p2.core.ProvisionException: No repository found at file:/D:/DevFiles/Eclipse/Tools/Dsl/net.certiv.dsl.jdt/net.certiv.dsl.jdt.site/target/repository.
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.fail (AbstractRepositoryManager.java:405)
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.loadRepository (AbstractRepositoryManager.java:709)
    at org.eclipse.equinox.internal.p2.metadata.repository.MetadataRepositoryManager.loadRepository (MetadataRepositoryManager.java:110)
    at org.eclipse.tycho.p2.remote.RemoteMetadataRepositoryManager.loadRepository (RemoteMetadataRepositoryManager.java:64)
    at org.eclipse.tycho.p2.remote.RemoteMetadataRepositoryManager.loadRepository (RemoteMetadataRepositoryManager.java:56)
    at org.eclipse.equinox.internal.p2.metadata.repository.CompositeMetadataRepository.addChild (CompositeMetadataRepository.java:171)
    at org.eclipse.equinox.internal.p2.metadata.repository.CompositeMetadataRepository.<init> (CompositeMetadataRepository.java:113)
    at org.eclipse.equinox.internal.p2.metadata.repository.CompositeMetadataRepositoryFactory.load (CompositeMetadataRepositoryFactory.java:124)
....

**Comment by laeubi** [comment]
I'm not sure where/what build is failing can you provide an example? The error itself is not from m2e so the best we can do is probably be more graceful in case of error reporting.

In general adding not existing repositories is an error and should be avoided even though it might work in some circumstance and better replaced by alternative techniques.

**Reply by grosenberg** [reply]
The build is of any standard tycho maven project whose POM includes a `<repository>` reference.  Build using M2E from `Run -> Maven build`

>The error itself is not from m2e 

You are right. Looks like it stems from Tycho, though I have yet to actually find the plugin/file that is the source of the failure in the Github repository for Tycho. (Should I move the discussion there?)

org.eclipse.equinox.internal.p2.metadata.repository.MetadataRepositoryManager.loadRepository (MetadataRepositoryManager.java:110)

> In general adding not existing repositories is an error and should be avoided even though it might work in some circumstance

Do not want to change the blocking ERROR as the default, but rather to be able to specify a configuration option that instead notes the problem as a non-blocking Warning or Info.  Of course, if non-blocking, a build would subsequently succeed or fail depending on whether the absent repository was consequential.

Kind of guessing that the old Sonatype bridge connector was just completely ignoring any missing repositories -- the same behavior as `Help->Install new software` applies when directed to the composite repository.

**Comment by laeubi** [comment]
> The build is of any standard tycho maven project whose POM includes a <repository> reference

Obviously not, Tycho builds use repository all the day long with `repository` but without an error, still a repository **must exits** there is no way for Tycho to ignore a missing repository. So if you run the build as described, the repository must exits.

> Kind of guessing that the old Sonatype bridge connector was just completely ignoring any missing repositories -

If you run the maven build as describe no connector can do anything, it is simply as if you execute the maven build on the console.

**Comment by grosenberg** [comment]
Feels like we are missing something in this discussion. 

Just to be certain, you understand what I am describing when I say [composite repository](https://wiki.eclipse.org/Equinox/p2/Composite_Repositories_(new)), right?

**Reply by laeubi** [reply]
I know what a composite repository is, still Tycho **is more strict** as Eclipse and treat missing childs as **error** while Eclipse silently **ignore** these.

All of this has nothing to do with m2e and even less with connectors... as Run As starts **a new (maven) process**, so technically a connector can not interact their.

You can set `-Declipse.p2.atomic.composite.loading.default=false` to disable that, but be aware that your build can become in-stable because P2 ignores *any* error ...

**Reply by grosenberg** [reply]
Adding `<property name='p2.atomic.composite.loading' value='false'/>` to the composite repository restores the desired non-blocking behavior. 

Thank you very much for providing the key to finding the solution!
